### PR TITLE
Build Native Debian installers consuming Portable Linux binaries

### PR DESCRIPTION
**Pull request by rakeshsinghranchi**
This work enables building debian installers ( Ubuntu 14.04/16.04/16.10  and Debian 8 ) consuming Linux Portable binaries.
Change involves : 
1. Updating existing build definition for Portable Linux and adding build steps to generate and publish installers for aforementioned distors and versions
2. Stop building native installers building native binaries for each aforementioned distro

Note: There is separate step to build docker image and then run the build for each distro version. There is a goal to consume the docker image from docker registry than building during the vso build step.

### CODE REVIEW

File: build_projects/dotnet-host-build/PackageTargets.cs [hunk 1]
```diff
@@ -266,5 +269,25 @@ private static void FixPermissions(string directory)
...
+
+            // set to default path if PORTABLE_BINARY_LOCATION environment variable doesn't exist
+            if(preBuiltPortableBinaryPath == null)
+            {
+                sharedFrameworkPublishPath = Dirs.SharedFrameworkPublish;
+            }
+            else 
+            {
+                Console.WriteLine($"Installers will package binaries from path set by PORTABLE_BINARY_LOCATION environment variable :{preBuiltPortableBinaryPath}");
+                sharedFrameworkPublishPath = preBuiltPortableBinaryPath;
```
Comment (rakeshsinghranchi) [inline build_projects/dotnet-host-build/PackageTargets.cs]: Is it okay to use Console.WriteLine() here or expected to call BuildContext.warn() ? I really want to add the log here.

File: build_projects/dotnet-host-build/PublishTargets.cs [hunk 2]
```diff
@@ -288,6 +298,13 @@ private static bool CheckIfAllBuildsHavePublished(BuildTargetContext c)
         public static BuildTargetResult PublishArtifacts(BuildTargetContext c) => c.Success();
 
         [Target(
+            nameof(PublishTargets.PublishInstallerFilesToAzure),
+            nameof(PublishTargets.PublishDotnetDebToolPackage),
+            nameof(PublishTargets.PublishDebFilesToDebianRepo),
+            nameof(PublishTargets.PublishSharedFrameworkVersionBadge))]
```
Comment (rakeshsinghranchi) [inline build_projects/dotnet-host-build/PublishTargets.cs]: One can use the Environment variable check to determine a target to be called or not.  I have followed this approach of grouping the applicable targets.

File: scripts/dockerbuild.sh [hunk 3]
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env bash
```
Comment (rakeshsinghranchi) [inline scripts/dockerbuild.sh]: Added a dockerbuild.sh file , future goal is to eliminate the step of building docker images during the build process,rather consume from docker registry. @gkhanna79 has suggested opening an issue related to this.

File: scripts/dockerbuild.sh [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
+  DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+  SOURCE="$(readlink "$SOURCE")"
+  [[ "$SOURCE" != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
+done
+DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
+
+cd "$DIR/.."
+
+INTERACTIVE="-i"
```
Comment (eerhardt) [inline scripts/dockerbuild.sh]: Where is this variable used?  I don't see it being used.

Comment (rakeshsinghranchi) [inline scripts/dockerbuild.sh]: Not used, will remove.

File: build_projects/dotnet-host-build/PackageTargets.cs [hunk 5]
```diff
@@ -266,5 +269,25 @@ private static void FixPermissions(string directory)
                 FS.FixModeFlags(directory);
             }
         }
+
+        private static string GetSharedFrameworkPublishPath()
```
Comment (chcosta) [inline build_projects/dotnet-host-build/PackageTargets.cs]: I guess it wasn't clear to me that you were making changes to the build system itself, when you said packaging, I interpreted that as the stuff under the pkg folder.  Most of this will be going away in the next week or two.  What's the priority on getting this done?  Would it make sense to wait on this, or to make parallel changes in my fork based on buildtools?

Comment (rakeshsinghranchi) [inline build_projects/dotnet-host-build/PackageTargets.cs]: CCing @gkhanna79 @Petermarcu  to provide more details regarding the priority . As of now, I am targeting 4/14 (EOW) to get this change in.

File: buildpipeline/Core-Setup-PortableLinux-x64.json [hunk 6]
```diff
@@ -97,8 +97,180 @@
...
+      "alwaysRun": false,
+      "displayName": "Copy built Portable binaries to staging directory",
+      "timeoutInMinutes": 0,
+      "task": {
+        "id": "5bfb729a-a7c8-4a78-a7c3-8d717bb7c13c",
+        "versionSpec": "2.*",
+        "definitionType": "task"
+      },
+      "inputs": {
+        "SourceFolder": "$(Build.SourcesDirectory)/artifacts/linux-x64/intermediate/sharedFrameworkPublish",
```
Comment (eerhardt) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Why are we doing this copy? Why not just point the PORTABLE_BINARY_LOCATION to here?

File: buildpipeline/Core-Setup-PortableLinux-x64.json [hunk 7]
```diff
@@ -97,8 +97,180 @@
...
+        "disableAutoCwd": "false",
+        "cwd": "",
+        "failOnStandardError": "false"
+      }
+    },
+    {
+      "enabled": true,
       "continueOnError": true,
-      "alwaysRun": true,
+      "alwaysRun": false,
```
Comment (eerhardt) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: ?? Is this good?  I think we've seen resources run out on the machines if we don't do this.

File: scripts/dockerbuild.sh [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+#  Jenkins
+[ ! -z "$BUILD_TAG" ] && DOTNET_BUILD_CONTAINER_NAME="${BUILD_TAG}-${DOCKER_TAG}"
+#  VSO
+[ ! -z "$BUILD_BUILDID" ] && DOTNET_BUILD_CONTAINER_NAME="${BUILD_BUILDID}-${BUILD_BUILDNUMBER}-${DOCKER_TAG}"
+
+# Executes a command and retries if it fails.
+# NOTE: This function is the exact copy from init-docker.sh.
+# Reason for not invoking init.docker.sh directly is since that script 
+# also performs cleanup, which we do not want in this case.
+execute() {
```
Comment (eerhardt) [inline scripts/dockerbuild.sh]: Is there any way to reuse this logic so it isn't copied with `dockerrun.sh`?

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 6 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Because 'git clean -xdf' is called for each step of building docker image. Git clean is required and removes the artifacts folder. Copying to staging directory to preserves the folder.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 7 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Thanks for pointing this out,will find out more details about alwaysRun usage and update the value accordingly.

File: build_projects/dotnet-host-build/PackageTargets.cs (hunk 5 above)
Comment (gkhanna79) [inline build_projects/dotnet-host-build/PackageTargets.cs]: Yes, we should get this in for 4/14 and any new infra would rebase against this to pick it up.

File: buildpipeline/Core-Setup-PortableLinux-x64.json [hunk 9]
```diff
@@ -97,8 +97,180 @@
...
+      "displayName": "Build Ubuntu.14.04 docker image",
+      "timeoutInMinutes": 0,
+      "task": {
+        "id": "6c731c3c-3c68-459a-a5c9-bde6e6595b5b",
+        "versionSpec": "2.*",
+        "definitionType": "task"
+      },
+      "inputs": {
+        "scriptPath": "scripts/dockerbuild.sh",
+        "args": "--non-interactive -d $(Build.SourcesDirectory)/scripts/docker/ubuntu.14.04 -t ubuntu.14.04-coresetup",
```
Comment (gkhanna79) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Can we use the Docker images @MichaelSimons has created instead of using the scripts to create one?

File: build_projects/dotnet-host-build/PackageTargets.cs [hunk 10]
```diff
@@ -266,5 +269,25 @@ private static void FixPermissions(string directory)
...
+        {
+            string sharedFrameworkPublishPath = string.Empty;
+            
+            string preBuiltPortableBinaryPath=Environment.GetEnvironmentVariable("PORTABLE_BINARY_LOCATION")?.Trim();
```
Comment (gkhanna79) [inline build_projects/dotnet-host-build/PackageTargets.cs]: This should be read in Prepare.Init and set within the build context and then you should use BuildContext to read the value here.

Comment (gkhanna79) [inline build_projects/dotnet-host-build/PackageTargets.cs]: Also, I would name the variable to "PORTABLE_BUILD_STAGING_LOCATION".

File: scripts/dockerbuild.sh (hunk 8 above)
Comment (MichaelSimons) [inline scripts/dockerbuild.sh]: I would think dockerrun.sh could call into this script to build and therefore no logic would need to be duplicated.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 6 above)
Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Can you explain why it is necessary to git clean -xdf in each build step?  This seems wrong.

---
In reply to: [110995126](https://github.com/dotnet/core-setup/pull/2027#discussion_r110995126) [](ancestors = 110995126)

File: buildpipeline/Core-Setup-PortableLinux-x64.json [hunk 11]
```diff
@@ -97,8 +97,180 @@
...
+      "displayName": "Build Debian Installer in Ubuntu.14.04 docker container",
+      "timeoutInMinutes": 0,
+      "task": {
+        "id": "6c731c3c-3c68-459a-a5c9-bde6e6595b5b",
+        "versionSpec": "2.*",
+        "definitionType": "task"
+      },
+      "inputs": {
+        "scriptPath": "scripts/dockerrun-as-current-user.sh",
+        "args": "--rm $(DockerBuildEnvironmentVariables) -e CLI_NUGET_API_KEY -e CLI_NUGET_FEED_URL -e CLI_NUGET_SYMBOLS_FEED_URL -v $(Build.SourcesDirectory):/opt/code/ -v $(Build.StagingDirectory)/sharedFrameworkPublish/:/opt/sharedFrameworkPublish/ -w /opt/code ubuntu.14.04-coresetup /bin/bash -c \"HOME=/opt/code; git clean -X -d -f; ./build.sh --skip-prereqs --configuration Release --targets $(InstallerOnlyBuildArguments)\"",
```
Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: I realize this is an existing pattern but is setting HOME necessary?  Does specifying the working directory achieve the same?

Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Why are the CLI_NUGET... variable needed only for ubuntu.14.04?

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 7 above)
Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: This should not be changed.  Same for the other cleanup tasks.

File: buildpipeline/Core-Setup-PortableLinux-x64.json [hunk 12]
```diff
@@ -222,6 +394,33 @@
...
+    "REPO_USER": {
+      "value": "dotnet"
+    },
+    "REPO_PASS": {
+      "value": "PassedViaPipeBuild"
+    },
+    "REPO_SERVER": {
+      "value": "azure-apt-cat.cloudapp.net"
+    },
+    "DockerBuildEnvironmentVariables": {
```
Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: The name is a little confusing since it is not used for the `Run build in Docker container` step

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 11 above)
Comment (eerhardt) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Typically NuGet starts throwing errors if $HOME isn't set.  That's why the pattern exists.

```
Unhandled Exception: System.ArgumentNullException: Value cannot be null.
13:17:17 Parameter name: path1
13:17:17    at System.IO.Path.Combine(String path1, String path2, String path3)
13:17:17    at NuGet.Common.NuGetEnvironment.GetFolderPath(SpecialFolder folder)
13:17:17    at NuGet.Common.NuGetEnvironment.GetFolderPath(NuGetFolderPath folder)
13:17:17    at NuGet.Configuration.SettingsUtility.GetHttpCacheFolder()
13:17:17    at NuGet.Configuration.NuGetPathContext.Create(ISettings settings)
```

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 9 above)
Comment (chcosta) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: +1 since this is something I've done in my refactoring but I had intended to run by @gkhanna79

Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: My understanding is that we want to get rid of building docker image on the fly as part of the build step. See issue : https://github.com/dotnet/core-setup/issues/1719  . It will be easier to remove this step and consume the prebuilt images in the build step. Once issue 1719 is resolved, we can simply remove this step.

File: build_projects/dotnet-host-build/PackageTargets.cs (hunk 10 above)
Comment (rakeshsinghranchi) [inline build_projects/dotnet-host-build/PackageTargets.cs]: Agreed,will make this change.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 11 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: This is required because PublishDebToolPackage Target requires it. see here : https://github.com/dotnet/core-setup/blob/master/build_projects/dotnet-host-build/PublishTargets.cs#L64

File: scripts/dockerbuild.sh (hunk 8 above)
Comment (rakeshsinghranchi) [inline scripts/dockerbuild.sh]: My understanding is that we want to get rid of building docker image on the fly as part of the build step. See issue : #1719 . It will be easier to remove this step and consume the prebuilt images from dockerhub in the build step. Once issue 1719 is resolved, we can simply remove this script.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 7 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Thanks Michael. will take care of this.

File: scripts/dockerbuild.sh (hunk 8 above)
Comment (MichaelSimons) [inline scripts/dockerbuild.sh]: While I agree 1719 is the best long term fix, the reality is it will probably be some time before that happens.  Duplicating this script logic is problematic in the interim.  Refactoring it now looks pretty easy.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 6 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: cc @gkhanna79  
if I don't call git clean step and artifacts are left behind from the previous build , I have noticed that coreclr fails to initialize in the prepare step. @gkhanna79  has more insights on why this happens.  The same approach of calling  git clean is done for portable Linux workflow https://github.com/dotnet/core-setup/blob/master/buildpipeline/Core-Setup-PortableLinux-x64.json#L92 here.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 9 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: @chcosta, @MichaelSimons  : do we have docker images that can be pulled from docker hub? My understanding is that #1719  is tracking the required work.

File: scripts/dockerbuild.sh (hunk 8 above)
Comment (rakeshsinghranchi) [inline scripts/dockerbuild.sh]: @MichaelSimons , are you suggesting updating dockerrun.sh to replace the docker build step at https://github.com/dotnet/core-setup/blob/master/scripts/dockerrun.sh#L127 with call to dockerbuild.sh . dockerbuild.sh generates its own custom tag and image name. Please help clarify the suggestion.

Comment (MichaelSimons) [inline scripts/dockerbuild.sh]: yes that is what I am suggesting.  Can't dockerrun.sh just `dockerbuild.sh -t $DOTNET_BUILD_CONTAINER_TAG -d $DOCKERFILE`.  e.g. pass in the custom tag is wants.

File: scripts/dockerbuild.sh [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+# This eliminates intermittent failures during docker build caused by failing to retrieve the base image.
+if [ ! -z "$image" ]; then
+    echo "Pulling Docker image $image"
+    execute docker pull $image
+fi
+
+docker build --build-arg USER_ID=$(id -u) -t $DOCKER_TAG $DOCKERFILE
+
+set +e
+docker rm -f $DOTNET_BUILD_CONTAINER_NAME
```
Comment (MichaelSimons) [inline scripts/dockerbuild.sh]: This like seems irrelevant here.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 9 above)
Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Yes we have docker images that can be used.  It is not clear to me what the requirements are for what the images contain.  The Dockerfiles are [here](https://devdiv.visualstudio.com/DevDiv/_git/DotNetCore?path=%2Fdockerfiles&version=GBmaster&_a=contents).  The images are [here](https://hub.docker.com/r/microsoft/dotnet-buildtools-prereqs/tags/).

File: scripts/dockerbuild.sh (hunk 13 above)
Comment (rakeshsinghranchi) [inline scripts/dockerbuild.sh]: do you mean Line# 97 and 98 ?

Comment (MichaelSimons) [inline scripts/dockerbuild.sh]: Line# 98 - $DOTNET_BUILD_CONTAINER_NAME does not appear in the script.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 9 above)
Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: I am looking for images equivalent to the docker file. for e.g. Ubuntu 14.04 as in https://github.com/dotnet/core-setup/blob/master/scripts/docker/ubuntu.14.04/Dockerfile  . Container will be used for building debian package.

Comment (rakeshsinghranchi) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Docker image for building debian packages are not available yet. Updated #1719 added reference to Dockerfile to required packages for building debian packages.

**Comment by rakeshsinghranchi** [comment]
@MichaelSimons @gkhanna79 @eerhardt  - All the feedbacks are addressed. Please review the changes.

File: build_projects/dotnet-host-build/PackageTargets.cs [hunk 14]
```diff
@@ -266,5 +269,25 @@ private static void FixPermissions(string directory)
...
             }
         }
+
+        private static string GetSharedFrameworkPublishPath(BuildTargetContext c)
+        {
+            string sharedFrameworkPublishPath = string.Empty;
+            
+            string preBuiltPortableStagingPath=c.BuildContext.Get<string>("PortableBuildStagingLocation");
+
+            // set to default path if PORTABLE_BUILD_STAGING_LOCATION environment variable doesn't exist
```
Comment (gkhanna79) [inline build_projects/dotnet-host-build/PackageTargets.cs]: I think you can update the comment to be more reflective of the intent - "If we are not generating distro specific installers for portable build, then we wont have access to staging location and thus, will use default   binary location where SharedFX was published".

File: scripts/dockerbuild.sh (hunk 8 above)
Comment (rakeshsinghranchi) [inline scripts/dockerbuild.sh]: Taken care of this.

File: buildpipeline/Core-Setup-PortableLinux-x64.json (hunk 12 above)
Comment (MichaelSimons) [inline buildpipeline/Core-Setup-PortableLinux-x64.json]: Any thoughts on this?

File: scripts/dockerbuild.sh (hunk 8 above)
Comment (MichaelSimons) [inline scripts/dockerbuild.sh]: Shouldn't the `set +e` have been kept?

**Comment by rakeshsinghranchi** [comment]
@dotnet-bot Test this please
